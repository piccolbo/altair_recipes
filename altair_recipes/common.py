"""Mini-library for all the other modules."""
import altair as alt
from boltons.iterutils import remap
from logging import warning
import numpy as np
import pandas as pd
import pytest_regtest
from toolz.dicttoolz import keyfilter, valfilter


def default(*args):
    """Return the first not None of the arguments.

    Parameters
    ----------
    *args : Any
        Any number of arguments.

    Returns
    -------
    Any
        One of the arguments.

    """
    return next(x for x in args if x is not None)


# testing
def viz_reg_test(test_f):
    """Decorate recipe tests.

    Transforms a function into a regression test. Also saves chart in html for
    visual inspection in the test directory, named after the file and function
    of the test. If invoked with None as sole argument, the decorated function
    will just produce the chart, disabling the regression machinery, again for
    manual inspection.

    Parameters
    ----------
    test_f : function
        Simple chart-generating argumentless function, pytest-style

    Returns
    -------
    function
        The decorated function.

    """

    def fun(regtest):
        with alt.data_transformers.enable(consolidate_datasets=False):
            np.random.seed(seed=0)
            plot = test_f()
            if regtest is not None:
                regtest.write(
                    alt.Chart.from_dict(round_floats(plot.to_dict(), 13)).to_json()
                )
                plot.save(
                    test_f.__code__.co_filename + "_" + test_f.__qualname__ + ".html"
                )
            return plot

    test_f.__doc__ = (
        (
            test_f.__doc__
            or "Test for function {test_f}".format(test_f=test_f.__qualname__)
        )
        + """
    Parameters
    ----------
    Pass a single unnamed argument equal None to manually  execute outside regression testing. In that case it returns a chart.
    """
    )
    return fun


@pytest_regtest.register_converter_pre
def fix_before(txt):
    """Remove schema information from regression data."""

    # remove lines with passwords:
    lines = txt.split('\n')
    lines = [l for l in lines if "https://vega.github.io/schema/vega-lite/" not in l]
    return '\n'.join(lines)

# collections


def check_distinct(data, col, group=None):
    """Check that a column of data contains unique values, optionally within each group defined by group.

    Parameters
    ----------
    data : dataframe
        The data.
    col : int or str or other pandas col
        The column of data to check.
    group : mapping, function, label, or list of labels
        optionally group by this and check uniqueness within each group (the default is None or do not group). See pandas.DataFrame.groupby for details.

    Returns
    -------
    bool
        True if all element distinct, for all groups within each group if group is not None.

    """
    if group is None:
        x = data[col]
        return x.size == x.nunique()
    else:
        x = data.groupby(group)[col]
        return all(x.size() == x.nunique())

def warn_not_distinct(data, col, group=None):
    """Generate warning if not distinct, see check_distinct."""

    if not check_distinct(data, col, group):
        warning("The relation to plot is not a function")


def choose_kwargs(from_, which):
    """Choose entries for a dictionary with key in `which` and not None value."""
    return keyfilter(lambda x: x in which, valfilter(lambda x: x is not None, from_))


def round_floats(a_dict, precision):
    """Find all the floats in `a_dict` (recursive) and round them to `precision`."""
    return remap(
        a_dict,
        lambda p, k, v: (k, round(v, precision)) if isinstance(v, float) else (k, v),
    )


def ndistinct(data, column):
    """Return number of distinct elements in data[column]."""
    return len(data[column].unique())


def col_cardinality(data, column, condition=None, default=1):
    """Return number of distinct elements in `data[column]` if `condition` is True (defaults to `column` being different from None), otherwise return `default`."""
    if condition is None:
        condition = column is not None
    return ndistinct(data, column) if condition else default


def gather(data, key, value, columns):
    """Convert wide format data frame to long format.

    Do so while concatenating selected columns into one and using a new column
    to track their origin.

    Parameters
    ----------
    data : pandas DataFrame
        The data to operate on.
    key : str
        The name of the column tracking the origin of that record.
    value : type
        The name of the column holding all the values previously in a number of
        columns.
    columns : list of str
        The names of the columns to reduce to a single column.

    Returns
    -------
    pandas.DataFrame
        A data frame with a reduced number of columns but the same information.

    """
    return pd.melt(
        data,
        id_vars=[col for col in data.columns if col not in columns],
        value_vars=columns,
        var_name=key,
        value_name=value,
    )


# TODO: this doesn't cover multiscatterplot which is a multivariate viz but does
# require the data in wide format, this only converts to long (gather)
# To include multiscatterplot one would need an index column or set thereof


def multivariate_preprocess(data, columns, group_by):
    """Preprocess data for multivariate graphs.

    Converts to data frame, then turns to long format.

    Parameters
    ----------
    data : pandas.DataFrame
        The data to be processed.
    columns : list of str
        The columns that need to be gathered.
    group_by : str
        The column indicating which vaiable a record refers to (long format).

    Returns
    -------
    (pandas.DataFrame, str, str)
        A tuple with the data in long format, the name of the column indicating
        the variable and the name of of the column holding the values.

    """
    assert (
        type(columns) == str or len(columns) == 1 or group_by is None
    ), "Wide or long format but not both"
    if group_by is None:  # convert wide to long
        key = default(data.columns.name, "variable")
        value = "value"
        data = gather(data, key=key, value=value, columns=columns)
    else:
        key = group_by
        value = columns if type(columns) is str else columns[0]
    return data, key, value


# constant luminosity and chroma scales

hue_scale_light = alt.Scale(type="linear", range=["#F271B8", "#00B4D7"])
hue_scale_dark = alt.Scale(type="linear", range=["#C10083", "#0080A7"])

# chart combinators


def layer(*layers, **kwargs):
    """Layer charts: a drop in replacement for altair.layer that does a deepcopy of the layers to avoid side-effects and lifts identical datasets one level down to top level."""
    layers = [l.copy() for l in layers]
    data = layers[0].data
    if all(map(lambda l: data.equals(l.data), layers)):
        layered = alt.layer(*layers, **kwargs, data=data)
        for l in layered.layer:
            del l._kwds["data"]
    else:
        layered = alt.layer(*layers, **kwargs)

    return layered
