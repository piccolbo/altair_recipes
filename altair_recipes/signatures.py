"""Collection of signatures used throughout the package."""

from .converters import Column, Columns, to_dataframe, init_cols
from altair import Chart, LayerChart
from autosig import Signature, param, Retval
from functools import partial

recipe = Signature(
    Retval(
        validator=lambda x: isinstance(x, (Chart, LayerChart)),
        docstring="""type altair.Chart or altair.LayerChart
    An altair Chart.""",
    ),
    data=param(
        default=None,
        converter=to_dataframe,
        position=0,
        docstring="""`altair.Data` or `pandas.DataFrame` or csv or json file URL
    The data from which the statistical graphics is being generated""",
    ),
    height=param(
        default=600,
        converter=int,
        position=-2,
        docstring="""`int`
    The height of the chart""",
    ),
    width=param(
        default=800,
        converter=int,
        position=-1,
        docstring="""`int`
    The width of the chart""",
    ),
).set_late_init(init_cols)

column = partial(param, converter=Column)

univariate_recipe = recipe + Signature(
    column=column(
        default=0,
        position=1,
        docstring="""`int`, `str`, pandas `Series` or a type convertible to it.
    The column containing the data to be used in the graphics""",
    )
)

bivariate_recipe = recipe + Signature(
    x=column(
        default=0,
        position=1,
        docstring="""`int`, `str`, pandas `Series` or a type convertible to it.
    The column containing the data associated with the horizontal dimension""",
    ),
    y=column(
        default=1,
        position=2,
        docstring="""`int`, `str`, pandas `Series` or a type convertible to it.
    The column containing the data associated with the vertical dimension""",
    ),
)

multivariate_recipe = recipe + Signature(
    columns=param(
        default=None,
        converter=Columns,
        position=1,
        docstring="""collection of: `int`, `str`, pandas `Series` or a type convertible to it.
    The column or columns to be used in the graphics, defaults to all""",
    ),
    group_by=column(
        default=None,
        position=2,
        docstring="""`int`, `str`, pandas `Series` or a type convertible to it.
    The column to be used to group the data when in long form. When group_by is
    specified columns should contain a single column""",
    ),
)

color = partial(
    column,
    docstring="""`str` or `int`
    The column containing the data associated with the color of the mark""",
)

use_color = partial(
    param,
    default=False,
    converter=bool,
    docstring="""bool
    Whether to also use color to encode the same data as the x coordinate""",
)

opacity = partial(
    param,
    default=1,
    converter=float,
    docstring="""float
    The value of the constant opacity of the mark (use to counter overlap)""",
)

tooltip = partial(
    column,
    docstring="""`str` or `int`
    The column containing the data associated with the tooltip text""",
)
