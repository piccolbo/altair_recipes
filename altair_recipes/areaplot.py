"""areaplots."""
from .common import warn_not_distinct, choose_kwargs
from .signatures import bivariate_recipe, color
import altair as alt
from autosig import autosig, Signature, param
from enum import Enum


class StackType(Enum):
    """Type of stacking for areaplot charts.

    Attributes
    ----------
    auto : None
        Automatic stacking.
    true : bool
        Stacked.
    false : bool
        Not stacked.
    normalize : string
        Normalize stacked to 1.

    """

    auto = None  # no, must not pass anything
    true = True
    false = False
    normalize = "normalize"


@autosig(
    bivariate_recipe
    + Signature(
        color=color(default=None, position=3),
        stack=param(
            default=StackType.auto,
            position=4,
            converter=StackType,
            docstring="""StackType
            One of `StackType.auto` (automatic selection), `StackType.true` (force), `StackType.false` (no stacking) and `StackType.normalize` (for normalized stacked)""",
        ),
    )
)
def areaplot(
    data=None, x=0, y=1, color=None, stack=StackType.auto, height=600, width=800
):
    """Generate an areaplot."""
    warn_not_distinct(data, x, color)
    if stack is not StackType.auto:
        y = alt.Y(y, stack=stack.value)
    opt_args = choose_kwargs(locals(), ["color"])
    return (
        alt.Chart(data=data, height=height, width=width)
        .mark_area(opacity=1)
        .encode(x=x, y=y, **opt_args)
    )
