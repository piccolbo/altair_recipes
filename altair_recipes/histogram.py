"""Generate histograms."""
from .common import multivariate_preprocess, col_cardinality
from .signatures import univariate_recipe, multivariate_recipe
import altair as alt
from autosig import autosig


@univariate_recipe
def histogram(data=None, column=0, height=600, width=800):
    """Generate a histogram."""
    return (
        alt.Chart(data, height=height, width=width)
        .mark_bar()
        .encode(alt.X(column, bin=True), alt.Y("count()"))
    )


@multivariate_recipe
def layered_histogram(data=None, columns=None, group_by=None, height=600, width=800):
    """Generate multiple overlapping histograms."""
    data, key, value = multivariate_preprocess(data, columns, group_by)
    return (
        alt.Chart(data, height=height, width=width)
        .mark_area(
            opacity=1 / col_cardinality(data, group_by, default=len(columns)),
            interpolate="step",
        )
        .encode(alt.X(value, bin=True), alt.Y("count()", stack=None), alt.Color(key))
    )
