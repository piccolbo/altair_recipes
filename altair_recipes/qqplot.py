"""Generate qqplots."""
from .signatures import bivariate_recipe
import altair as alt
from autosig import autosig
import numpy as np


@bivariate_recipe
def qqplot(data=None, x=0, y=1, height=600, width=800):
    """Generate a quantile-quantile plot."""
    n = data.shape[0]
    data = data.quantile(np.arange(0, 1, step=10 / n))
    return alt.Chart(data, height=height, width=width).mark_point().encode(x=x, y=y)
