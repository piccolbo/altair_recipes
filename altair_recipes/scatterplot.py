"""Scatterplots."""
from .common import choose_kwargs, hue_scale_dark, hue_scale_light
from .signatures import bivariate_recipe, multivariate_recipe, color, tooltip
import altair as alt
from autosig import autosig, Signature, param
from numbers import Number

scatterplot_sig = Signature(
    color=color(default=None, position=3),
    opacity=param(
        default=1,
        position=4,
        converter=float,
        docstring="""`float`
    A constant value for the opacity of the mark""",
    ),
    tooltip=tooltip(default=None, position=5),
)


@autosig(bivariate_recipe + scatterplot_sig)
def scatterplot(
    data=None, x=0, y=1, color=None, opacity=1, tooltip=None, height=600, width=800
):
    """Generate a scatterplot."""
    if color is not None:
        if isinstance(data[color].iloc[0], Number):
            color = alt.Color(
                color, scale=hue_scale_light if opacity == 1 else hue_scale_dark
            )
    opt_args = choose_kwargs(locals(), ["color", "tooltip"])
    return alt.Chart(
        data,
        height=height,
        width=width,
        mark=alt.MarkDef(type="point" if opacity == 1 else "circle", opacity=opacity),
    ).encode(x=x, y=y, **opt_args)


@autosig(multivariate_recipe + scatterplot_sig)
def multiscatterplot(
    data=None,
    columns=None,
    group_by=None,
    color=None,
    opacity=1,
    tooltip=None,
    height=600,
    width=800,
):
    """Generate many scatterplots.

    Based on several columns, pairwise.
    """
    opt_args = choose_kwargs(locals(), ["color", "tooltip"])

    assert group_by is None, "Long format not supported yet"
    return (
        alt.Chart(data, height=height // len(columns), width=width // len(columns))
        .mark_point(size=1 / len(columns), opacity=opacity)
        .encode(
            alt.X(alt.repeat("column"), type="quantitative"),
            alt.Y(alt.repeat("row"), type="quantitative"),
            **opt_args
        )
        .repeat(row=columns, column=columns)
    )
