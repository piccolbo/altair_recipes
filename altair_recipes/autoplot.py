"""Automatic plot selection."""
from .barchart import barchart
from .boxplot import boxplot
from .common import col_cardinality
from .heatmap import heatmap
from .histogram import histogram
from .scatterplot import scatterplot
from .signatures import multivariate_recipe
from .stripplot import stripplot
from autosig import autosig
from numbers import Number
import pandas as pd


def bin_midpoints(xx, n):
    """Divide a vector in equal-width bins and return the midpoints if it is Numeric, otherwise the vector itself.

    Parameters
    ----------
    xx : Iterable
        The vector to be binned or other iterable to return as-is.
    n : integer
        Number of bins.

    Returns
    -------
    list or same as xx
        The midpoints or xx itself.

    """
    return (
        list(map(lambda x: x.mid, list(pd.cut(xx, n))))
        if issubclass(type(xx.iloc[0]), Number)
        else xx
    )


def overlap(data):
    """Find how dense a set of points is in relation to plotting them.

    Parameters
    ----------
    data : DataFrame
        The data whose density needs to be evaluated.

    Returns
    -------
    integer
        The maximum number of data points having identical coordinates on categorical dimensions or falling within the same bin when Numeric, after an equal-width binning operation into 100 bins (100 is somewhat arbitrary, experience based).

    """
    return data.apply(bin_midpoints, n=100).groupby(list(data.columns)).size().max()


def is_cat(xx):
    """Whether a list or vector is categorical (non Numeric).

    Parameters
    ----------
    xx : list or vector
        A list or vector.

    Returns
    -------
    bool
        Whether `xx` is categorical.

    """
    return not issubclass(type(xx.iloc[0]), Number)


@multivariate_recipe
def autoplot(data=None, columns=None, group_by=None, height=600, width=800):
    """Automatically choose and produce a statistical graphics based on up to three columns of data."""

    assert group_by is None, "long data not supported yet"
    vars_n = len(columns)
    assert vars_n <= 3, "Only up to three vars supported at this time"
    data = data[columns]
    columns = sorted(columns, key=lambda x: -len(data[x].unique()))
    y, x, z, *_ = columns + 2 * [None]
    overlap_deg = overlap(data[columns[: min(vars_n, 2)]])
    max_overlap = 10
    high_overlap = overlap_deg >= max_overlap
    no_overlap = overlap_deg == 1
    low_overlap = 1 < overlap_deg < max_overlap
    cat_vars_n = sum(map(lambda col: is_cat(data[col]), columns))
    numeric_vars_n = vars_n - cat_vars_n

    chart_type_selection = [
        (scatterplot, numeric_vars_n >= 2 and not high_overlap),
        (heatmap, numeric_vars_n >= 2 and high_overlap),
        (stripplot, numeric_vars_n == 1 and not high_overlap),
        (barchart, numeric_vars_n == 0),
        (histogram, numeric_vars_n == 1 and cat_vars_n == 0 and high_overlap),
        (boxplot, numeric_vars_n == 1 and cat_vars_n >= 1 and high_overlap),
    ]
    chart_type = list(filter(lambda x: x[1], chart_type_selection))
    assert len(chart_type) == 1
    chart_type = chart_type[0][0]
    use_facet = cat_vars_n >= 2 or (
        cat_vars_n == 1 and numeric_vars_n == 2 and not no_overlap
    )
    use_color = vars_n == 3 and not (use_facet and chart_type is scatterplot)
    use_opacity = (chart_type in (scatterplot, stripplot) and low_overlap) or (
        chart_type is heatmap and high_overlap and numeric_vars_n == 3
    )  # heat

    if chart_type is barchart:
        args = dict(x=y, y="count()", height=height, width=width)
        if use_facet:
            args.update(x=x, color=x, width=width // col_cardinality(data, y))
            facet_args = dict(column=y)
            if z is not None:
                args.update(height=height // col_cardinality(data, z))
                facet_args.update(row=z)
        chart = barchart(data, **args)
        if use_facet:
            chart = chart.facet(**facet_args)

    if chart_type is boxplot:
        chart = boxplot(
            data,
            columns=y,
            group_by=x,
            color=use_facet,  # this is a redundant use of color, different from necessary uses. Encodes x again
            height=height,
            width=width // col_cardinality(data, z),
        )
        if use_facet:
            chart = chart.facet(column=z)
    if chart_type is heatmap:
        args = dict(color="count()")
        if use_opacity:
            args.update(color=z, opacity="count()")
        chart = heatmap(
            data,
            x=x,
            y=y,
            aggregate="average" if use_opacity else "count",
            height=height // col_cardinality(data, z, use_facet),
            width=width // col_cardinality(data, z, use_facet),
            **args
        )
        if use_facet:
            chart = chart.facet(row=z)
    if chart_type is histogram:
        chart = histogram(data, column=y, height=height, width=width)
    if chart_type is stripplot:
        chart = stripplot(
            data,
            columns=y,
            group_by=x,
            color=x if use_facet else None,
            opacity=1 / overlap_deg if use_opacity else 1,
            height=height,
            width=width // col_cardinality(data, z, use_facet),
        )
        if use_facet:
            chart = chart.facet(column=z)
    if chart_type is scatterplot:
        chart = scatterplot(
            data,
            x=x,
            y=y,
            color=z if use_color else None,
            opacity=1 / overlap_deg if use_opacity else 1,
            height=height // col_cardinality(data, z, use_facet),
            width=width // col_cardinality(data, z, use_facet),
        )
        if use_facet:
            chart = chart.facet(row=z)
    return chart
