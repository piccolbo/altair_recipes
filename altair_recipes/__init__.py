# -*- coding: utf-8 -*-
"""Top-level package for altair_recipes."""
from .autocorrelation import autocorrelation
from .autoplot import autoplot
from .barchart import barchart
from .boxplot import boxplot
from .common import layer
from .heatmap import heatmap
from .histogram import histogram, layered_histogram
from .lineplot import lineplot
from .areaplot import areaplot, StackType
from .qqplot import qqplot
from .scatterplot import scatterplot, multiscatterplot
from .smoother import smoother
from .stripplot import stripplot


# alphabetically ordered exports
__all__ = [
    "areaplot",
    "autocorrelation",
    "autoplot",
    "barchart",
    "boxplot",
    "layer",
    "lineplot",
    "heatmap",
    "histogram",
    "layered_histogram",
    "multiscatterplot",
    "qqplot",
    "scatterplot",
    "smoother",
    "StackType",
    "stripplot",
]
__author__ = """Antonio Piccolboni"""
__email__ = "altair_recipes@piccolboni.info"
__version__ = "__version__ = '0.9.0'"
