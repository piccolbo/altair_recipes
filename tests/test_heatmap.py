import altair_recipes as ar
from altair_recipes.common import viz_reg_test
from altair_recipes.display_pweave import show_test
import numpy as np
import pandas as pd
from vega_datasets import data


#' <h2>Heatmap</h2>


@viz_reg_test
def test_heatmap():
    # Compute x^2 + y^2 across a 2D grid
    x, y = np.meshgrid(range(-5, 6), range(-5, 6))
    z = x ** 2 + y ** 2

    # Convert this grid to columnar data expected by Altair
    data = pd.DataFrame({"x": x.ravel(), "y": y.ravel(), "z": z.ravel()})

    return ar.heatmap(data, x="x", y="y", color="z")


show_test(test_heatmap)


#' <h2>Count Heatmap</h2>


@viz_reg_test
def test_count_heatmap():
    source = data.movies.url
    return ar.heatmap(
        source, x="IMDB_Rating", y="Rotten_Tomatoes_Rating", color="", aggregate="count"
    )


show_test(test_count_heatmap)
