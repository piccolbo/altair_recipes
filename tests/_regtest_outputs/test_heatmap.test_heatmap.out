{
  "$schema": "https://vega.github.io/schema/vega-lite/v4.8.1.json",
  "config": {
    "view": {
      "continuousHeight": 300,
      "continuousWidth": 400
    }
  },
  "data": {
    "values": [
      {
        "x": -5,
        "y": -5,
        "z": 50
      },
      {
        "x": -4,
        "y": -5,
        "z": 41
      },
      {
        "x": -3,
        "y": -5,
        "z": 34
      },
      {
        "x": -2,
        "y": -5,
        "z": 29
      },
      {
        "x": -1,
        "y": -5,
        "z": 26
      },
      {
        "x": 0,
        "y": -5,
        "z": 25
      },
      {
        "x": 1,
        "y": -5,
        "z": 26
      },
      {
        "x": 2,
        "y": -5,
        "z": 29
      },
      {
        "x": 3,
        "y": -5,
        "z": 34
      },
      {
        "x": 4,
        "y": -5,
        "z": 41
      },
      {
        "x": 5,
        "y": -5,
        "z": 50
      },
      {
        "x": -5,
        "y": -4,
        "z": 41
      },
      {
        "x": -4,
        "y": -4,
        "z": 32
      },
      {
        "x": -3,
        "y": -4,
        "z": 25
      },
      {
        "x": -2,
        "y": -4,
        "z": 20
      },
      {
        "x": -1,
        "y": -4,
        "z": 17
      },
      {
        "x": 0,
        "y": -4,
        "z": 16
      },
      {
        "x": 1,
        "y": -4,
        "z": 17
      },
      {
        "x": 2,
        "y": -4,
        "z": 20
      },
      {
        "x": 3,
        "y": -4,
        "z": 25
      },
      {
        "x": 4,
        "y": -4,
        "z": 32
      },
      {
        "x": 5,
        "y": -4,
        "z": 41
      },
      {
        "x": -5,
        "y": -3,
        "z": 34
      },
      {
        "x": -4,
        "y": -3,
        "z": 25
      },
      {
        "x": -3,
        "y": -3,
        "z": 18
      },
      {
        "x": -2,
        "y": -3,
        "z": 13
      },
      {
        "x": -1,
        "y": -3,
        "z": 10
      },
      {
        "x": 0,
        "y": -3,
        "z": 9
      },
      {
        "x": 1,
        "y": -3,
        "z": 10
      },
      {
        "x": 2,
        "y": -3,
        "z": 13
      },
      {
        "x": 3,
        "y": -3,
        "z": 18
      },
      {
        "x": 4,
        "y": -3,
        "z": 25
      },
      {
        "x": 5,
        "y": -3,
        "z": 34
      },
      {
        "x": -5,
        "y": -2,
        "z": 29
      },
      {
        "x": -4,
        "y": -2,
        "z": 20
      },
      {
        "x": -3,
        "y": -2,
        "z": 13
      },
      {
        "x": -2,
        "y": -2,
        "z": 8
      },
      {
        "x": -1,
        "y": -2,
        "z": 5
      },
      {
        "x": 0,
        "y": -2,
        "z": 4
      },
      {
        "x": 1,
        "y": -2,
        "z": 5
      },
      {
        "x": 2,
        "y": -2,
        "z": 8
      },
      {
        "x": 3,
        "y": -2,
        "z": 13
      },
      {
        "x": 4,
        "y": -2,
        "z": 20
      },
      {
        "x": 5,
        "y": -2,
        "z": 29
      },
      {
        "x": -5,
        "y": -1,
        "z": 26
      },
      {
        "x": -4,
        "y": -1,
        "z": 17
      },
      {
        "x": -3,
        "y": -1,
        "z": 10
      },
      {
        "x": -2,
        "y": -1,
        "z": 5
      },
      {
        "x": -1,
        "y": -1,
        "z": 2
      },
      {
        "x": 0,
        "y": -1,
        "z": 1
      },
      {
        "x": 1,
        "y": -1,
        "z": 2
      },
      {
        "x": 2,
        "y": -1,
        "z": 5
      },
      {
        "x": 3,
        "y": -1,
        "z": 10
      },
      {
        "x": 4,
        "y": -1,
        "z": 17
      },
      {
        "x": 5,
        "y": -1,
        "z": 26
      },
      {
        "x": -5,
        "y": 0,
        "z": 25
      },
      {
        "x": -4,
        "y": 0,
        "z": 16
      },
      {
        "x": -3,
        "y": 0,
        "z": 9
      },
      {
        "x": -2,
        "y": 0,
        "z": 4
      },
      {
        "x": -1,
        "y": 0,
        "z": 1
      },
      {
        "x": 0,
        "y": 0,
        "z": 0
      },
      {
        "x": 1,
        "y": 0,
        "z": 1
      },
      {
        "x": 2,
        "y": 0,
        "z": 4
      },
      {
        "x": 3,
        "y": 0,
        "z": 9
      },
      {
        "x": 4,
        "y": 0,
        "z": 16
      },
      {
        "x": 5,
        "y": 0,
        "z": 25
      },
      {
        "x": -5,
        "y": 1,
        "z": 26
      },
      {
        "x": -4,
        "y": 1,
        "z": 17
      },
      {
        "x": -3,
        "y": 1,
        "z": 10
      },
      {
        "x": -2,
        "y": 1,
        "z": 5
      },
      {
        "x": -1,
        "y": 1,
        "z": 2
      },
      {
        "x": 0,
        "y": 1,
        "z": 1
      },
      {
        "x": 1,
        "y": 1,
        "z": 2
      },
      {
        "x": 2,
        "y": 1,
        "z": 5
      },
      {
        "x": 3,
        "y": 1,
        "z": 10
      },
      {
        "x": 4,
        "y": 1,
        "z": 17
      },
      {
        "x": 5,
        "y": 1,
        "z": 26
      },
      {
        "x": -5,
        "y": 2,
        "z": 29
      },
      {
        "x": -4,
        "y": 2,
        "z": 20
      },
      {
        "x": -3,
        "y": 2,
        "z": 13
      },
      {
        "x": -2,
        "y": 2,
        "z": 8
      },
      {
        "x": -1,
        "y": 2,
        "z": 5
      },
      {
        "x": 0,
        "y": 2,
        "z": 4
      },
      {
        "x": 1,
        "y": 2,
        "z": 5
      },
      {
        "x": 2,
        "y": 2,
        "z": 8
      },
      {
        "x": 3,
        "y": 2,
        "z": 13
      },
      {
        "x": 4,
        "y": 2,
        "z": 20
      },
      {
        "x": 5,
        "y": 2,
        "z": 29
      },
      {
        "x": -5,
        "y": 3,
        "z": 34
      },
      {
        "x": -4,
        "y": 3,
        "z": 25
      },
      {
        "x": -3,
        "y": 3,
        "z": 18
      },
      {
        "x": -2,
        "y": 3,
        "z": 13
      },
      {
        "x": -1,
        "y": 3,
        "z": 10
      },
      {
        "x": 0,
        "y": 3,
        "z": 9
      },
      {
        "x": 1,
        "y": 3,
        "z": 10
      },
      {
        "x": 2,
        "y": 3,
        "z": 13
      },
      {
        "x": 3,
        "y": 3,
        "z": 18
      },
      {
        "x": 4,
        "y": 3,
        "z": 25
      },
      {
        "x": 5,
        "y": 3,
        "z": 34
      },
      {
        "x": -5,
        "y": 4,
        "z": 41
      },
      {
        "x": -4,
        "y": 4,
        "z": 32
      },
      {
        "x": -3,
        "y": 4,
        "z": 25
      },
      {
        "x": -2,
        "y": 4,
        "z": 20
      },
      {
        "x": -1,
        "y": 4,
        "z": 17
      },
      {
        "x": 0,
        "y": 4,
        "z": 16
      },
      {
        "x": 1,
        "y": 4,
        "z": 17
      },
      {
        "x": 2,
        "y": 4,
        "z": 20
      },
      {
        "x": 3,
        "y": 4,
        "z": 25
      },
      {
        "x": 4,
        "y": 4,
        "z": 32
      },
      {
        "x": 5,
        "y": 4,
        "z": 41
      },
      {
        "x": -5,
        "y": 5,
        "z": 50
      },
      {
        "x": -4,
        "y": 5,
        "z": 41
      },
      {
        "x": -3,
        "y": 5,
        "z": 34
      },
      {
        "x": -2,
        "y": 5,
        "z": 29
      },
      {
        "x": -1,
        "y": 5,
        "z": 26
      },
      {
        "x": 0,
        "y": 5,
        "z": 25
      },
      {
        "x": 1,
        "y": 5,
        "z": 26
      },
      {
        "x": 2,
        "y": 5,
        "z": 29
      },
      {
        "x": 3,
        "y": 5,
        "z": 34
      },
      {
        "x": 4,
        "y": 5,
        "z": 41
      },
      {
        "x": 5,
        "y": 5,
        "z": 50
      }
    ]
  },
  "encoding": {
    "color": {
      "aggregate": "average",
      "field": "z",
      "scale": {
        "range": [
          "#F271B8",
          "#00B4D7"
        ],
        "type": "linear"
      },
      "type": "quantitative"
    },
    "x": {
      "bin": {
        "maxbins": 14.0
      },
      "field": "x",
      "type": "quantitative"
    },
    "y": {
      "bin": {
        "maxbins": 8.0
      },
      "field": "y",
      "type": "quantitative"
    }
  },
  "height": 600,
  "mark": "rect",
  "width": 800
}